File: 0078-subsets/0078-subsets.java
class Solution {
    public void dfs(List<List<Integer>> answer, List<Integer> container, int[] nums, int index){
        answer.add(new ArrayList(container)); 
        for(int i = index; i < nums.length; i++){
            container.add(nums[i]);
            dfs(answer, container, nums, i + 1);
            container.remove(container.size()-1); 
        }
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> answer = new ArrayList(); 
        List<Integer> container = new ArrayList<>(); 
        dfs(answer, container, nums, 0); 
        return answer; 
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void dfs(vector<int>& nums, vector<vector<int>>& answer, vector<int>& container, int index){
        answer.push_back(container); 
        for(int i = index; i < nums.size(); i++){
            container.push_back(nums[i]); 
            dfs(nums,answer,container,i+1);
            container.pop_back(); 
        }
    }
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> answer; 
        vector<int> container; 
        dfs(nums,answer,container,0); 
        return answer; 
    }
};